New depletion acceptance rule

jglaser opened the pull request:
## Description

This is the first PR of a series to introduce a new acceptance algorithm for depletion interactions. The new algorithm uses a Metropolis Hastings rule based on an estimate of the free energy change. The previous algorithm, while observing detailed balance, was based on rejection sampling in the overlap volume. One advantage is that the new estimator does not require a fixed average density of test particle insertions, this can now be chosen arbitrarily, and the estimated free energies are rescaled accordingly. E.g., for low sampling density the estimator *variance* is high, leading to a large number of rejected configurations, whereas a high sampling density is computationally more expensive but affords a higher acceptance rate.

To ensure correct sampling, the free energy estimates become part of the state probability itself and therefore the configuration of depletants around particles needs to be tracked as part of the configuration. Luckily, we can save memory by just storing the random seed used to generate the depletants per particle in one additional variable, here the .x component of the velocity, which is otherwise unused in MC. This strategy is inspired by the "pseudo-marginal" approach of [Andrieu and Roberts](https://projecteuclid.org/euclid.aos/1236693147).

The parameter controlling the estimator variance is called "ntrial" and integer. However, more generally, it can be any positive scalar and we'll change the definition in a subsequent PR. Since this is the first PR in a series, some other changes introduced here are also not yet final and will be reverted later (see below).

**New features**

1. Introduce a new *ad-hoc* MPI decomposition mode for depletants, along two axes (number of particles and number of depletants per particle). This parallel layer sits *in between* node-local GPU parallelism and spatial domain decomposition. It is enabled by passing a dedicated `mpi4py` communicator to specific methods of `hpmc.integrate_*`. The communicator is usually split from `MPI.COMM_WORLD`.

2. Make `update.clusters()` rejection-free for anisotropic particles (but no longer ergodic). In 3D, It only generates pi-rotations parallel to the box axes, and therefore does not sample all particle orientations; it needs to be combined with local moves. In 2D, it is also ergodic (pi rotations in plane).

4. Introduce full support for depletants in `UpdaterClusters`, and enable `update.clusters()`on the GPU. Rather than estimating the free energy change as described above, in this rejection-free algorithm depletants are treated as actual particles but efficiently generated on the fly.

5. **Temporary change**: Refactor the acceptance algorithm in HPMC on the GPU. The GPU algorithm parallelizes over the particles to compute overlaps and interactions, but we finalize the acceptance decision *in the individual interaction kernels* per particle, rather than deferring it to the acceptance kernel. This eliminates the need for temporary storage (neighbor lists), however it means that the interactions have to be recomputed in every iteration of the acceptance loop. 

6. **Temporary change** Introduce depletant type pairs instead of a single depletant type. This allows the solvent-solute interaction to be non-additive as well. The two particle types refer to the types of the two particles a depletant particle overlaps simultaneously to induce the depletion interaction. Upon further examination, this implementation appears to violate the n-body depletion potential and will be removed in a subsequent PR.

7. **Temporary change**  Introduce data structures to sample in the leaf nodes of union shapes for depletion (on the CPU)

**Incomplete features**:

- Implement patch interactions in `UpdaterClustersGPU`(will be tackled later)
- [ ] Either reinstate `ntrial=0` depletants feature on the CPU (by copy and paste from the earlier branch), or remove it (also on the GPU) **EDIT** will remove in next PR

**Other changes**:

- Improve build system by automatically generating the template instantiations for GPU kernels through CMake, and place individual kernels in single .cu files. This partially addresses #614, to resolve that issue, this implementation can serve as a template for the `module_*.cc` files for the CPU classes. I am deferring this change to later in the v3 development process to avoid too many merge conflicts.
- Remove vestigial quermass code
- `update.clusters()` no longer supports spatial domain decomposition.
- Port `update.muvt()` to new API

## Motivation and Context

In the long term, we'd like to implement a fast model for implicit solvation in an aqueous solvent. This PR generalizes the acceptance rule to eventually take into account enthalpic terms in the solute-solvent interaction, and potentially, also solvent pair correlations (at least, if those correspond to 'bosonic' systems, which depletants are.... but that is another story)

## How Has This Been Tested?

Validation and unit tests have been amended as necessary.

## Change log

New features:
- New Metropolis acceptance rule for depletants that relies on free energy estimates with tunable variance (ntrial parameter)
- `hpmc.update.Clusters()` now supports the GPU and depletants
- HPMC: particles and depletants can be split across GPUs using MPI with or without spatial domain decompsoition (Python API to be completed)

Other changes:
- `hpmc.update.Clusters()` is now rejection free, but applies pi-rotations only along box axes and therefore is no longer ergodic (needs to be combined with single particle moves)
- `hpmc.update.Clusters()` does no longer support spatial domain decomposition with MPI 

## Checklist:

- [x] I have reviewed the [**Contributor Guidelines**](https://github.com/glotzerlab/hoomd-blue/blob/master/CONTRIBUTING.md).
- [x] I agree with the terms of the [**HOOMD-blue Contributor Agreement**](https://github.com/glotzerlab/hoomd-blue/blob/master/ContributorAgreement.md).
- [x] My name is on the [list of contributors](https://github.com/glotzerlab/hoomd-blue/blob/master/sphinx-doc/credits.rst).

joaander commented:
Thanks, Jens. I'm won't to be able to fully review this until after my vacation next week.

In the meantime, it would be helpful if you could split of some of the smaller bugfixes and more isolated changes into their own small PRs. @lyrivera is going to do this for the LLVM 10 support commits. At least that will isolate the massive PR to the more structural changes to HPMC.

jglaser commented:
OK, also thank you @lyrivera .

I will try my best to go through it once more by myself to hopefully split off smaller changes.

Have a great vacation!

jglaser commented:
Will complete work in this branch soon, hopefully I can get in before the "big merge conflict" (https://github.com/glotzerlab/hoomd-blue/tree/feature/new-object-API)

joaander commented:
Jens, at this point we are all putting our effort into getting v3.0-beta.1 complete and will merge feature/new-object-API into master at that time. We don't have any bandwidth to review this PR until after that, so we'd appreciate it if you could merge the new API into this branch.

There is still quite a bit of work to do before beta 1: https://github.com/glotzerlab/hoomd-blue/projects/2 - but we hope to finish it very soon.

joaander commented:
Let us know when you have this ready to review based on master. In the meantime we are continuing with a number of large projects that will create more merge conflicts.

jglaser commented:
> Let us know when you have this ready to review based on master. In the meantime we are continuing with a number of large projects that will create more merge conflicts.

I am starting to merge in master this coming week, hoping to have this first PR (of a series) ready for review soon.

jglaser commented:
OK, I am trying to make my way through the failing unit tests. Looks like all of them pass but one:

```    raise NotImplementedError("map_energies will be implemented in a future"
                                  "release.")
E       NotImplementedError: map_energies will be implemented in a futurerelease.

hoomd/hpmc/integrate.py:228: NotImplementedError
=========================== short test summary info ============================
SKIPPED [20] hoomd/conftest.py:210: Validation tests not requested.
SKIPPED [1] hoomd/md/pytest/test_pair.py:656: Cannot test force energy relationship for DPD pair force
SKIPPED [1] hoomd/md/pytest/test_pair.py:656: Cannot test force energy relationship for DPDLJ pair force
SKIPPED [344] hoomd/conftest.py:163: Test is run only on GPU(s).
FAILED hoomd/hpmc/pytest/test_write_debug_data_hpmc.py::test_write_debug_data[CPU]
================= 1 failed, 1957 passed, 366 skipped in 45.88s =================
```

It looks like the `NotImplementedError` should also be triggered on `master`, so I don't know how to proceed. I can keep working on adding some extra tests and run them locally, but CI will only complete once this issue is resolved.

How are validation tests handled? Using pytest? The old way? Do I need to convert the depletion test?

joaander commented:
> It looks like the `NotImplementedError` should also be triggered on `master`, so I don't know how to proceed. I can keep working on adding some extra tests and run them locally, but CI will only complete once this issue is resolved.

I'm not aware of any tests using `map_energies` on master. We haven't implemented `map_energies` yet. If you need this functionality, feel free to open a separate focused PR on master.

> 
> How are validation tests handled? Using pytest? The old way? Do I need to convert the depletion test?

We are using pytest. Mark validation tests with `@pytest.mark.validate` and use `pytest.mark.parameterize` and/or fixtures to parameterize tests. Validation tests do not run by default: See https://hoomd-blue.readthedocs.io/en/latest/testing.html#running-validation-tests for docs on how to trigger them. CI will run validation tests on all branches at ~10pm daily.

We have not ported all the old v2 validation tests yet as we are currently prioritizing getting functionality working. Feel free to port the test in this PR or leave it for later. If not, we will port it later when we get to the other validation tests. If you do implement the depletion test, please get it to run in as short a time as possible.

jglaser commented:
> > It looks like the `NotImplementedError` should also be triggered on `master`, so I don't know how to proceed. I can keep working on adding some extra tests and run them locally, but CI will only complete once this issue is resolved.
> 
> I'm not aware of any tests using `map_energies` on master. We haven't implemented `map_energies` yet. If you need this functionality, feel free to open a separate focused PR on master.

Ah, right. There was a merge conflict which caused `map_energies` to have a `@log` decorator, which means it was being considered automatically in some write debug data unit test. Fixed.

jglaser commented:
Basic tests are passing. At this point, code review can begin. Thanks. And let me know if you're daunted by the amount of changes. I might be able to separate certain parts, like the clusters ones, into their own PR. But it will be more work for me, and they will have to be reviewed eventually anyway. Let me know how you'd like to proceed.

joaander commented on hoomd/hpmc/CMakeLists.txt:
```diff
@@ -73,38 +97,67 @@ set(_hpmc_headers
 
+if(ENABLE_HIP)
+    # enable out-of-source build
+    target_include_directories(_hpmc PUBLIC "$<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}>")
```
The source directory may not be available once hoomd is installed, or the source directory may have headers that are not compatible with the installation.  This should be `PRIVATE` so that plugins don't attempt to include files from the source directory.
```suggestion
    target_include_directories(_hpmc PRIVATE "$<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}>")
```

jglaser replied:
done (sorry, I could have just committed your suggestion).

joaander commented on hoomd/hpmc/GPUHelpers.cuh:
```diff
@@ -33,12 +45,14 @@ __device__ inline unsigned int computeParticleCell(const Scalar3& p,
 
     // identify the bin
-    if (f.x >= Scalar(0.0) && f.x < Scalar(1.0) && f.y >= Scalar(0.0) && f.y < Scalar(1.0) && f.z >= Scalar(0.0) && f.z < Scalar(1.0))
+    if (!strict || (f.x >= Scalar(0.0) && f.x < Scalar(1.0) &&
```
What does this accomplish? When strict is false, out of bounds ib, jb, or kb values will result in returns values wrapping to other cells in the list and possibly cells outside of the addressable range.

jglaser replied:
`computeParticleCell()` is now used in more than one place, in the move proposal kernel as well as in the overlap kernels. 
See here https://github.com/glotzerlab/hoomd-blue/blob/770703b87e8f0a5aa4e6c083124cd12764325192/hoomd/hpmc/IntegratorHPMCMonoGPUMoves.cuh#L110-L111 and here https://github.com/glotzerlab/hoomd-blue/blob/770703b87e8f0a5aa4e6c083124cd12764325192/hoomd/hpmc/IntegratorHPMCMonoGPUMoves.cuh#L149-L150
In these examples, the `strict=true` argument is passed, as opposed to the places where it's being used in the overlap kernels, e.g. in
https://github.com/glotzerlab/hoomd-blue/blob/770703b87e8f0a5aa4e6c083124cd12764325192/hoomd/hpmc/IntegratorHPMCMonoGPUDepletants.cuh#L204-L205

The reason for the distinction is that an particle that is out of the box generates a return value for the bin (`0xffffffff`) that is always out of bounds, a condition that can therefore easily be intercepted, as above. The moves kernel then sets the `d_reject_out_of_cell` flag for that particle, which is checked by the overlaps kernel and if set, allows it to skip any out of bounds memory accesses to the cell list. The `false` argument is not really necessary (passing true would be perfectly fine), it's intended to be more of a statement for anyone who reads the code that at that point we rely on the cell being valid.

joaander replied:
Alright, that makes sense. The place where `false` is passed doesn't mind if it gets an in-bounds value for a particle that is out of the box.

joaander commented on hoomd/hpmc/HPMCCounters.h:
```diff
@@ -388,6 +343,8 @@ DEVICE inline hpmc_implicit_counters_t operator+(const hpmc_implicit_counters_t&
     hpmc_implicit_counters_t result;
     result.insert_count = a.insert_count + b.insert_count;
+    result.insert_accept_count = a.insert_accept_count + b.insert_accept_count;
+    result.insert_accept_count_sq = a.insert_accept_count_sq + b.insert_accept_count_sq;
```
Squared counts do not add and subtract. Why do you need to store squared counts in addition to counts?

jglaser replied:
This was an attempt at calculating the variance of the energy estimator. It will go away in the next PR.

tommy-waltmann replied:
Why not just remove it from this PR then?

joaander replied:
@tommy-waltmann in the interests of keeping the PRs smaller we are going to merge some changes only to undo them later.  I didn't realize that this was one of those changes. I'm fine with adding it now - the beta releases are unstable.

joaander commented on hoomd/jit/CMakeLists.txt:
```diff
@@ -42,6 +42,8 @@ pybind11_add_module(_${PACKAGE_NAME} SHARED ${_${PACKAGE_NAME}_sources} ${_${PAC
 add_library(HOOMD::_${PACKAGE_NAME} ALIAS _${PACKAGE_NAME})
 
 add_library (_${PACKAGE_NAME}_llvm SHARED ${_${PACKAGE_NAME}_llvm_sources})
+set_property(TARGET _${PACKAGE_NAME}_llvm PROPERTY CXX_STANDARD 14)
```
This is no longer necessary. We now set a project wide `CMAKE_CXX_STANDARD 14`.

jglaser replied:
eliminated

joaander commented on hoomd/hpmc/update.py:
```diff
@@ -521,71 +407,56 @@ class Clusters(Updater):
-        delta_mu (float): The chemical potential difference between types to be swapped
-        move_ratio (float): Set the ratio between pivot and reflection moves.
-        flip_probability (float): Set the probability for transforming an
+        move_ratio(float): Set the ratio between pivot and reflection moves.
```
@vyasr pointed out that `move_ratio` parameters should be renamed #777.

jglaser replied:
done for clusters updater

joaander commented on hoomd/hpmc/update.py:
```diff
@@ -280,15 +280,16 @@ def perturb(timestep):
-        period (int): Number of timesteps between histogram evaluations.
-        transfer_types (list): List of type names that are being transferred from/to the reservoir or between boxes (if *None*, all types)
+        trigger (int): Number of timesteps between grand canonical insertions
+        transfer_types (list): List of type names that are being transferred from/to the reservoir or between boxes
```
Arguments that are also attributes should be listed again in an `Attributes:` doc section at the end of the class docstring.

jglaser replied:
adding partial documentation for attributes

[ ] make sure `update.muvt()` works, expose C++ setter/getters as attributes, and document

joaander commented on hoomd/hpmc/integrate.py:
```diff
@@ -133,16 +134,23 @@ def __init__(self, seed, d, a, translation_move_probability, nselect):
                                            type_kind='particle_types',
                                            param_dict=TypeParameterDict(
-                                               0., len_keys=1))
+                                               0., len_keys=2)
```
Update the documentation for `depletant_fugacity` to state that it takes two types and explain meaning of fugacity as a type matrix.

jglaser replied:
I clarified as suggested that the feature is temporary and will be removed in the subsequent PR, where the fugacity will be defined per type as before. As explained above, the type pair option was meant to support depletants that behave as particle type `A` when overlapping with another particle, and as particle type `B` when overlapping with the particle that particle is overlapping with. This treatment is not consistent with the many body character of the depletion interaction and is only correct when there are no higher order than pair interactions in the system.

joaander replied:
Understood. The betas are unstable and its OK if we make and unmake breaking changes. I may delay merging this until after the beta 4 release hoping that the next PR could be merged before beta 5.

joaander commented on hoomd/hpmc/integrate.py:
```diff
@@ -133,16 +134,23 @@ def __init__(self, seed, d, a, translation_move_probability, nselect):
+                                               0., len_keys=2)
+                                           )
+
+        typeparam_ntrial = TypeParameter('depletant_ntrial',
```
Add documentation for `depletant_ntrial` to the class docstring.

jglaser replied:
done

joaander reviewed (changes requested):
Thanks Jens! I have a few requested changes. I didn't review the details of the new depletant code, I'll assume that you implemented and tested it well. I'm looking forward to getting this and the upcoming PRs merged so we can eventually move forward on the hipper conversion.

Additionally, we'll also need to write change log entries (as a PR comment). In addition the the Clusters changes and new features I noticed that you added support for MuVT and modified the depletant parameters in the HPMC integrators.

tommy-waltmann commented on hoomd/CachedAllocator.h:
```diff
@@ -236,18 +216,7 @@ T* CachedAllocator::getTemporaryBuffer(size_t num_elements)
 //                << "reached; removing unused block ("
 //                << float(i->first)/1024.0f/1024.0f << " MB)" << std::endl;
 
-            // transform the pointer to cuda::pointer before calling cuda::free
-            #ifdef __HIP_PLATFORM_HCC__
-            if (m_managed)
-                {
-                // workaround, HIP doesn't fully support managed memory API
-                hipHostFree((void *) i->second);
-                }
-            else
-            #endif
-                {
-                hipFree((void *) i->second);
-                }
```
There are a few places where a change like this is made. Can you explain why this change is necessary?

jglaser replied:
in HIP on AMD platforms, managed memory is being emulated as zero copy memory and in the early days of HIP (around version 2.9), actually required one to use the `hipHostFree` API to free managed memory instead of `hipFree`. Now at least there is API parity with CUDA on that aspect.

joaander commented:
@jglaser let me know when you think you can get to addressing this review. I'm concerned that completing #894 and #640 will create many more merge conflicts with this branch.

jglaser commented:
working on it this week

jglaser commented on hoomd/MPIConfiguration.h:
```diff
@@ -45,6 +45,11 @@ class PYBIND11_EXPORT MPIConfiguration
         virtual ~MPIConfiguration() {};
 
 #ifdef ENABLE_MPI
+        MPI_Comm operator()() const
```
I realized that `MPIConfiguration` is really just a holder for an MPI communicator (and the world communicator), therefore I thought it was appropriate to expose the wrapped communicator via an interface that is slightly less verbose than the explicit `getCommunicator()`.  See https://github.com/glotzerlab/hoomd-blue/blob/f7ffceb46104490b3feda0278923930713741327/hoomd/hpmc/IntegratorHPMCMonoGPU.h#L771-L775 for an example.

I also believe the name `Configuration` suggests slightly more than what it is and could be changed to `MPICommunicator` going forward.

joaander replied:
Certainly. Feel free to open an issue to note the needed changes and/or a separate PR.

jglaser replied:
done #915

jglaser commented on hoomd/extern/ECL.cuh:
```diff
@@ -0,0 +1,430 @@
+implementation thereof is very fast. It operates on graphs stored in
+binary CSR format.
+
+Copyright (c) 2017, Texas State University. All rights reserved.
```
done

jglaser commented:
Thanks, @joaander and @tommy-waltmann . I addressed the comments. As last steps, I need to verify that the validation tests are passing and that the muvt updater is functional, too. Then we can move on to the next PR.

jglaser commented:
fixing the unit tests again, it remains to port the depletants validation test to pytest, and make sure `update.muvt()` works
- [ ] implement depletion validation test w/pytest

b-butler commented on hoomd/hpmc/update.py:
```diff
@@ -280,15 +280,16 @@ def perturb(timestep):
+        trigger (int): Number of timesteps between grand canonical insertions
+        transfer_types (list): List of type names that are being transferred from/to the reservoir or between boxes
         ngibbs (int): The number of partitions to use in Gibbs ensemble simulations (if == 1, perform grand canonical muVT)
+        max_volume_rescale (flaot): maximum step size in ln(V) (applies to Gibbs ensemble)
```
```suggestion
        max_volume_rescale (float): maximum step size in ln(V) (applies to Gibbs ensemble)
```

b-butler commented on hoomd/hpmc/update.py:
```diff
@@ -300,160 +301,49 @@ class muvt(_updater):
-        """
-        self.check_initialization();
+        param_dict = ParameterDict(seed=int(seed),
+                                   transfer_types=list(transfer_types),
```
`None` is not iterable and will cause an error here. In general, it is better to create the `ParameterDict` and then use update to set its values here what you want for the specification is `[str]`

jglaser replied:
good catch. I hadn't tested the muvt updater yet. Will fix when I test it.

jglaser replied:
unit test added.

b-butler commented on hoomd/hpmc/update.py:
```diff
@@ -300,160 +301,49 @@ class muvt(_updater):
-            muvt.set_params(dV=0.1)
-            muvt.set_params(n_trial=2)
-            muvt.set_params(volume_move_probability=0.05)
+        self.ngibbs = int(ngibbs)
```
Is there a reason that `ngibbs` isn't in the `param_dict`?

jglaser replied:
yes, it doesn't really make sense to dynamically reconfigure the MPI partition layout during execution..

b-butler commented on hoomd/hpmc/update.py:
```diff
@@ -631,67 +506,31 @@ def _attach(self):
 
-        Returns:
-            A tuple of (accepted moves, rejected moves) since the last run.
+        0.0 when not attached
```
We typically default to `None` when not attached.

jglaser replied:
fixed

b-butler commented on hoomd/hpmc/update.py:
```diff
@@ -631,67 +506,31 @@ def _attach(self):
-            return self._cpp_obj.getCounters(1)
-
-    @log(category='sequence')
-    def pivot_moves(self):
```
Why did we remove all move counters from the user facing API?

jglaser replied:
See above, since update.clusters() is now rejection-free there is no point in reporting an acceptance rate

b-butler commented on hoomd/hpmc/validation/lj_spheres.py:
```diff
@@ -109,7 +109,10 @@ def accumulate_energy(timestep):
         # Sample
-        run(1000,callback=accumulate_energy, callback_period=10)
+        if use_clusters:
+            run(5000,callback=accumulate_energy, callback_period=10)
```
Do these old `run` calls still work? This is no longer the syntax used for version 3 simulations.

jglaser replied:
No, they won't work. We need to port over the validation tests from 2 to 3 in a separate PR.

b-butler commented on hoomd/hpmc/pytest/test_clusters.py:
```diff
@@ -155,57 +137,10 @@ def test_valid_setattr_attached(attr, value, simulation_factory,
-
-@pytest.mark.parametrize("delta_mu", [-2.0, 0, 2.0])
-def test_pivot_moves(delta_mu, simulation_factory,
+@pytest.mark.serial
```
Does `Cluster` no longer work in MPI simulations?

jglaser replied:
No, it won't work. I removed the domain decomposition support (which was complicated and inefficient anyway). Cluster moves are global, so they don't map well onto a spatial domain decomposition.

b-butler reviewed (changes requested):
I have looked at all the Python code and some of the C++/HIP code though not thoroughly. I have left some comments to address or respond to.

jglaser commented on hoomd/RandomNumbers.h:
```diff
@@ -22,10 +22,10 @@
 #pragma GCC diagnostic push
 #pragma GCC diagnostic ignored "-Wconversion"
 #include <hoomd/extern/random123/include/Random123/philox.h>
+#include <type_traits>
```
see here 

https://github.com/glotzerlab/hoomd-blue/blob/b1228a668f1d7751d1ee3a0f3ddda8ff935c0f59/hoomd/RandomNumbers.h#L29-L30

and https://en.cppreference.com/w/cpp/types/make_signed

b-butler replied:
Thanks! I don't know why I missed these.

jglaser commented on hoomd/hpmc/kernel_cluster_depletants.cu.in:
```diff
@@ -0,0 +1,34 @@
+#cmakedefine SHAPE_INCLUDE @SHAPE_INCLUDE@ // the name of the include file
+#cmakedefine IS_UNION_SHAPE  // define to generate a kernel for a ShapeUnion<...>
+
+#define XSTR(x) #x
+#define STR(x) XSTR(x)
```
No, `#` has a special meaning in the C++ preprocessor. See here https://gcc.gnu.org/onlinedocs/cpp/Stringizing.html

b-butler replied:
Thanks!

jglaser commented on CMake/hoomd/HOOMDCUDASetup.cmake:
```diff
@@ -24,7 +24,7 @@ if (ENABLE_HIP)
         if (_cuda_min_arch LESS 60)
             message(SEND_ERROR "HOOMD requires compute 6.0 or newer")
-        endif()
+        endif ()
```
No

jglaser commented:
I added a basic unit test for depletion. Implementing the validation test will have to wait until we have ported `compute.free_volume()` to v3.

jglaser commented:
MuVT updater fixed and unit test added. Now need to make sure all tests pass and docs render OK, then this should be good to go.

vyasr commented on hoomd/hpmc/IntegratorHPMCMonoGPU.h:
```diff
@@ -186,49 +232,56 @@ class IntegratorHPMCMonoGPU : public IntegratorHPMCMono<Shape>
 
+        std::vector<hipStream_t> m_narrow_phase_streams;             //!< Stream for narrow phase kernel, per device
+        std::vector<std::vector<hipStream_t> > m_depletant_streams;  //!< Stream for every particle type, and device
+        std::vector<std::vector<hipStream_t> > m_depletant_streams_phase1;  //!< Streams for phase1 kernel
```
Could you document somewhere that the 2-phase process is specifically for when ntrial > 0? There's a lot of code here and it's hard to keep track of, and will be more so for anyone who's unfamiliar with the sampling process to begin with.

jglaser replied:
I've added a comment to that extent

https://github.com/glotzerlab/hoomd-blue/blob/1d0c644f74b3922dad8dd59a64305d6266142c97/hoomd/hpmc/IntegratorHPMCMonoGPU.h#L268

However, ntrial = 0 will (likely) be eliminated in subsequent PRs to avoid having two depletants code paths

vyasr reviewed (changes requested):
I reviewed this as thoroughly as I could. I didn't really look at the muvt updater or the GPU implementation of the clusters updater (the CPU implementation looks fine), but I did look through the new depletant code. There's a ton of code here and I'm sure I missed some things, but the overall logic seems sound and the sampling scheme for depletants looks OK, although I did have one or two questions. Very interested to know how this performs.

jglaser commented on hoomd/CommunicatorGPU.cc:
```diff
@@ -1490,8 +1490,6 @@ void CommunicatorGPU::migrateParticles()
         uint3 mypos = m_decomposition->getGridPos();
 
-        /* We need some better heuristics to decide whether to take the GPU or CPU code path */
-        #if 0
```
> Very interested to know how this performs.

Thanks, @vyasr!

Without detailed performance measurements I can just say it performs *decently*.  However, note that this is not the final iteration of this algorithm (more changes forthcoming in the next PR, as also noted above). One main change in the acceptance rule in this PR is to propose trial moves, and then to realize every temporary accept/reject decision per particle and re-check overlaps, *for each iteration of the acceptance kernel sequence*. This obviously represents an opportunity for optimization if we are able to pre-calculate the free energy change as a function of those decisions, rather than re-evaluating them several times (typically 3-4 times). As a consequence, this present implementation is a little slower than it could be. But this kind of logic was easier to implement in a first attempt with the two phase depletant insertion kernels.

In the next PR, the pre-calculation of overlaps will be (or, in fact, has is already been :-) reverted to the previous behavior of checking old and new configurations in one go, for improved performance. This also explains the change in the shift factors (`<< 2`) you correctly noted below, as those shifts were tracking the old and the new configuration of the neighboring particle.

vyasr replied:
Got it, so some of this code involves known performance regressions to simplify the 1-phase to 2-phase conversion.

jglaser replied:
> It appears that this code path was never safe before, but now the CPU fallback has been completely removed. What changed?

See below (comment on CommunicatorGPU.cu)

jglaser commented on hoomd/hpmc/IntegratorHPMCMonoGPU.h:
```diff
@@ -769,303 +1115,467 @@ void IntegratorHPMCMonoGPU< Shape >::update(unsigned int timestep)
+                                    CHECK_CUDA_ERROR();
+                                m_tuner_depletants_phase1->end();
+
+                                // phase 2, reinsert into excluded volume of i's neighbors
```
It looks like you're trying to understand the method. And to thoroughly explain it, I admit I really have to write a paper, otherwise from just looking at the code it may seem to come out of nowhere :-)

Let me try to answer your question. The idea of the algorithm is that a Poisson point process provides an unbiased estimator of the Boltzmann weight for the overlap volume, and every realization of such a process represents a microstate, the ensemble of which we sample through the usual MCMC (Markov Chain Monte Carlo).

To understand the need for the two kernels, note that if we're moving particle i, we are also proposing a new configuration of test particles overlapping with it  (stored compactly as a single random number generator seed in the auxiliary variable). The configuration of this particle and its test particles then represent the new microstate, and we compute the Metropolis-Hastings (MH) weight as the usual ratio of estimators (**phase 1**). Because other particles `j!=i` and their associated depletants remain in their existing configurations, we have to account for the possibility that i's move changes the outcome of its neighbor j's estimators. In particular, if i overlaps with any of j's test particles before or after the move, we have to include the new and the old estimate for j in both the numerator and the denominator of the acceptance probability, by re-inserting test particles for j-neighbors whenever we move i (**phase 2**).

Looking at the serial implementation you'll notice that we perform this reinsertion of test particles for neighbors with every trial move  - the number of depletant insertions therefore scales as `O(N_neigh)*O(N)`, which is slower than the `O(N)` of the simple rejection algorithm (in addition to the factor of two for evaluating numerator and denominator separately). Of course, there are still `N_avg` test particles being inserted per actual particle. The split into two kernels reduces the number of reinsertions back to `O(N)*N_avg` (on the GPU), by inserting all i test particles in phase 1, and by simultaneously inserting into all j-neighbors in phase 2 (the phases do not depend on each other). More specifically, in phase 2 kernel a j particle touches i's memory, where as in phase 1 the mapping between particles and acceptance probabilities is the identity. The slight implication of this access pattern (for multi-GPU performance) is that the random access from j -> i in phase 2 necessitates page migrations, if particles i and j are on different GPUs.

vyasr replied:
Yes, a paper would be nice :) but your logic makes sense, I see what you're doing. Thanks for the explanation!

joaander commented on hoomd/hpmc/.gitignore:
```diff
@@ -0,0 +1 @@
+kernel_*.cu
```
The current version of this branch puts the generated `kernel_*.cu` files in the build directory. We can just remove this from .gitignore.

jglaser commented on hoomd/CommunicatorGPU.cu:
```diff
@@ -187,9 +187,6 @@ void gpu_sort_migrating_particles(const unsigned int nsend,
     assert(d_end);
     assert(d_neighbors);
 
-    if (!nsend)
-        return;
-
```
The early exit was the reason for the buggy behavior, when there were 0 particles to send the output buffers and counts remained uninitialized but they are assumed to be initialized.

jglaser commented on hoomd/hpmc/IntegratorHPMCMono.h:
```diff
@@ -152,59 +154,79 @@ class IntegratorHPMCMono : public IntegratorHPMC
-        Scalar getDepletantFugacityPy(std::string type_name)
+        //! Set the depletant density in the free volume
+        void setDepletantFugacityPy(std::pair<std::string, std::string> types, Scalar fugacity)
             {
-            unsigned int id = this->m_pdata->getTypeByName(type_name);
-            return m_fugacity[id];
+            unsigned int type_a = this->m_pdata->getTypeByName(types.first);
+            unsigned int type_b = this->m_pdata->getTypeByName(types.second);
+            m_fugacity[m_depletant_idx(type_a,type_b)] = fugacity;
+            m_fugacity[m_depletant_idx(type_b,type_a)] = fugacity;
```
they are bound through the `TypeParameter` framework, see here

https://github.com/glotzerlab/hoomd-blue/blob/fdf9aa947234c50f592597f3853535eaac6613f2/hoomd/hpmc/integrate.py#L147

jglaser commented on hoomd/hpmc/IntegratorHPMCMonoGPU.h:
```diff
@@ -550,23 +754,85 @@ void IntegratorHPMCMonoGPU< Shape >::update(unsigned int timestep)
+        int particle_comm_rank;
+        if (m_particle_comm)
+            {
+            // split local particle data further if a communicator is supplied
```
yes, exactly.. it's just an optional API for (potentially) speeding up simulations with thousands or more depletants per particle that become slow on a single GPU

jglaser commented on hoomd/hpmc/IntegratorHPMCMonoGPUDepletants.cuh:
```diff
@@ -0,0 +1,838 @@
+#ifdef __HIP_PLATFORM_NVCC__
+__launch_bounds__(max_threads)
+#endif
+__global__ void hpmc_insert_depletants(const Scalar4 *d_trial_postype,
```
Usually the `__syncthreads` are not a problem by themselves, as long as the execution is load-balanced. That said, removing any of them would make the code incorrect.

jglaser commented on hoomd/hpmc/IntegratorHPMCMono.h:
```diff
@@ -319,7 +347,27 @@ class IntegratorHPMCMono : public IntegratorHPMC
+                o << " diameter charge";
+                }
+
+            bool have_auxilliary_variables = false;
```
I have fixed that in a subsequent branch... so in order to avoid merge conflicts I'd like to leave it as is in this one.

jglaser commented on hoomd/hpmc/IntegratorHPMCMonoGPUDepletants.cuh:
```diff
@@ -0,0 +1,838 @@
+        __syncthreads();
+
+        // process the queue, group by group
+        if (master && group == 0)
+            s_adding_depletants = 0;
+
+        if (master && group == 0)
+            {
```
OK

jglaser commented:
I addressed the comments, sorry for the delay. I hope we can move forward now - do you want a few lines for the ChangeLog?

joaander commented:
Yes, please write lines for the change log and post them as a PR comment or edit the description.

I'm going to merge #894 into master first - it is done. Then we will merge #934 into this branch which effectively merges master up into this branch. Could you check b20d1c62d60df00411989524f6db5a65bd87470a first to ensure I didn't break anything with the new seeding constructors?

Once that is done, we can merge this branch and move on to the next phase.

vyasr reviewed (approved):
@jglaser I applied some of my outstanding suggestions from my PR review. There are still a couple of change requests outstanding that I think would help clean up the code a little bit, so take a quick look through those before finalizing if you can.

jglaser commented on hoomd/hpmc/OBB.h:
```diff
@@ -77,7 +78,7 @@ struct OBB
 
     //! Default construct a 0 OBB
-    DEVICE OBB() : mask(1), is_sphere(0) {}
+    DEVICE OBB() : mask(0xffffffffu), is_sphere(0) {}
```
done

jglaser commented on hoomd/hpmc/OBB.h:
```diff
@@ -140,56 +141,192 @@ struct OBB
+        OverlapReal e, f;
+        e = M.row0.x*lower_a.x;
+        f = M.row0.x*upper_a.x;
+        if (e < f)
```
fixed.

jglaser commented on hoomd/hpmc/IntegratorHPMCMonoGPU.h:
```diff
@@ -260,22 +312,52 @@ IntegratorHPMCMonoGPU< Shape >::IntegratorHPMCMonoGPU(std::shared_ptr<SystemDefi
+            for (auto t: Autotuner::getTppListPow2(warp_size))
+                {
+                // only widen the parallelism if the shape supports it
+                if (t == 1 || Shape::isParallel())
```
yeah, but I don't see how it would lead to simpler or less code.

jglaser commented on hoomd/hpmc/IntegratorHPMCMonoGPU.h:
```diff
@@ -550,23 +754,85 @@ void IntegratorHPMCMonoGPU< Shape >::update(unsigned int timestep)
+        #endif
+
+        GPUPartition gpu_partition_rank = this->m_pdata->getGPUPartition();
+        unsigned int nparticles_rank = this->m_pdata->getN();
```
fixed

jglaser commented:
> @jglaser I applied some of my outstanding suggestions from my PR review. There are still a couple of change requests outstanding that I think would help clean up the code a little bit, so take a quick look through those before finalizing if you can.

Thanks, sorry that I missed those because they were hidden in the fold.

joaander reviewed (approved):
Thanks! Lets merge this and move on to the next phase.

@b-butler have your change requests been addressed?

jglaser commented:
Thank you, @joaander  for pushing this despite my long reply times. I have more code lined up, but it would be possible to target 3.1, too, depending on the timeline you're thinking of.

joaander commented:
I don't have a hard timeline in mind for 3.0. There still a lot of work to be done before everything is final and stabilized. https://github.com/glotzerlab/hoomd-blue/projects/3

How sweeping are the changes you have for the next phase? We have a few more sweeping changes to the codebase. We need to refactor all of the TBB for loops to remove deprecated TBB APIs. We also need to finish the hipper conversion. Would you recommend holding off on these or getting them started now?

jglaser commented:
> I don't have a hard timeline in mind for 3.0. There still a lot of work to be done before everything is final and stabilized. https://github.com/glotzerlab/hoomd-blue/projects/3
> 
> How sweeping are the changes you have for the next phase? We have a few more sweeping changes to the codebase. We need to refactor all of the TBB for loops to remove deprecated TBB APIs. We also need to finish the hipper conversion. Would you recommend holding off on these or getting them started now?

The changes are mostly localized to the HPMC (GPU) integrator and the Autotuner, with possible reorganization of the way the HPMC CPU code files are compiled (similar to the GPU ones in this PR). I will merge master into the subsequent branches, prepare PRs, and then we can decide their priority based on the size of the changes. The hipper changes are on my TODO.

joaander commented:
> The changes are mostly localized to the HPMC (GPU) integrator and the Autotuner, with possible reorganization of the way the HPMC CPU code files are compiled (similar to the GPU ones in this PR). I will merge master into the subsequent branches, prepare PRs, and then we can decide their priority based on the size of the changes. The hipper changes are on my TODO.

OK, so changes to the TBB loops shouldn't conflict, I'll get started on that. I'm looking forward to getting the hipper changes in.
